api/types/resource.go

```diff
@@ -704,3 +704,51 @@ func GetKind(v any) string {
 
 	return ""
 }
+
+// GetRevision returns the revision if one can be obtained.
+func GetRevision(v any) string {
```

Comment:
My first instinct was "hey, why don't we just extend the `ResourceMetadata` interface", but then I realized we'd have to extend protobuf messages with our own methods, and that might be even uglier. I guess there's no other way, right?

Reply:
This is doable, and can be pretty neat, but we would have to implement our own codegen for this. I believe custom codegen for resource types is much better route than what we currently have, but it would need some convincing before we could go that way.

Reply:
Custom codegen is a suggestion that often comes up. I'll admit I'm one of those that need convincing. I think there's plenty we could do to "cleanup" the codebase so there is less need to match our resources to various large and demanding interfaces.

It could also be that I'm just failing to visualize what folks have in mind.

Reply:
I find it that our code for resources is extremely repetitive. For example, let's assume a Foo resource:

```proto
message Foo {
  // common fields
  string kind = 1;
  string sub_kind = 2;
  string version = 3;
  teleport.header.v1.Metadata metadata = 4;

  // static spec
  FooSpec spec = 5;

  // dynamic status
  FooStatus status = 6;
}

message FooSpec {
  repeated string bars = 1;
}

message FooStatus {
  bool foo_bar_baz = 1;
}
```

I've omitted most comments for brevity, but this is more or less how a resource should look like (one with a status; most don't need one, so this can be refined even more). 

Now - to be able to actually use that resource? You need to add so much boilerplate code for creating, retrieving, deleting by name, deleting all, ... The next day a new resource comes along, similar to the previous one, but distinct:

```proto
message Bar {
  // common fields
  string kind = 1;
  string sub_kind = 2;
  string version = 3;
  teleport.header.v1.Metadata metadata = 4;

  // static spec
  BarSpec spec = 5;

  // dynamic status
  BarStatus status = 6;
}

message BarSpec {
  repeated string foos = 1;
}

message FooStatus {
  bool bar_foo_baz = 1;
}
```

You need to add code which is almost the same, yet not. This is the code that could be largely autogenerated. There would surely be need for extensibility of some kind, but this can be arranged.